File: bench/set_partition/set_partition.cpp
#include <algorithm>
#include <chrono>
#include <iostream>
#include <vector>

#define vec std::vector

namespace {
vec<int> range(int start, int stop) {
  vec<int> v(stop - start);
  uint j = 0;
  for (int i = start; i < stop; i++)
    v[j++] = i;
  return v;
}

bool conforms(const vec<vec<int>> &candidate, int minsize, int forgive) {
  int deficit = 0;
  for (const auto &p : candidate) {
    int need = minsize - p.size();
    if (need > 0)
      deficit += need;
  }
  return deficit <= forgive;
}

void partition_filtered(const vec<int> &collection,
                        std::function<void(const vec<vec<int>> &)> callback,
                        int minsize = 1, int forgive = 0) {
  if (collection.size() == 1) {
    callback({collection});
    return;
  }

  auto first = collection[0];
  auto loop = [&](const vec<vec<int>> &smaller) {
    int n = 0;
    for (const auto &subset : smaller) {
      vec<vec<int>> candidate;
      candidate.reserve(smaller.size());
      for (int i = 0; i < n; i++)
        candidate.push_back(smaller[i]);

      vec<int> rep;
      rep.reserve(subset.size() + 1);
      rep.push_back(first);
      rep.insert(rep.end(), subset.begin(), subset.end());
      candidate.push_back({rep});

      for (int i = n + 1; i < smaller.size(); i++)
        candidate.push_back(smaller[i]);

      if (conforms(candidate, minsize, forgive))
        callback(candidate);
      ++n;
    }

    vec<vec<int>> candidate;
    candidate.reserve(smaller.size() + 1);
    candidate.push_back({first});
    candidate.insert(candidate.end(), smaller.begin(), smaller.end());

    if (conforms(candidate, minsize, forgive))
      callback(candidate);
  };

  vec<int> new_collection(collection.begin() + 1, collection.end());
  partition_filtered(new_collection, loop, minsize, forgive + 1);
}
} // namespace

int main(int argc, char *argv[]) {
  using clock = std::chrono::high_resolution_clock;
  using std::chrono::duration_cast;
  using std::chrono::milliseconds;

  auto t = clock::now();
  int n = 1;
  int x = 0;

  auto callback = [&](const vec<vec<int>> &p) {
    auto copy = p;
    std::sort(copy.begin(), copy.end());
    x += copy[copy.size() / 3][0];
  };

  auto something = range(1, std::atoi(argv[1]));
  partition_filtered(something, callback, 2);
  std::cout << x << std::endl;
  std::cout << (duration_cast<milliseconds>(clock::now() - t).count() / 1e3)
            << std::endl;
}

File: bench/set_partition/set_partition.py
# https://stackoverflow.com/questions/73473074/speed-up-set-partition-generation-by-skipping-ones-with-subsets-smaller-or-large
import sys
import time

def conforms(candidate, minsize, forgive):
    """
    Check if partition `candidate` is at most `forgive` additions from making
    all its elements conform to having minimum size `minsize`
    """
    deficit = 0
    for p in candidate:
        need = minsize - len(p)
        if need > 0:
            deficit += need

    # Is the deficit small enough?
    return (deficit <= forgive)

def partition_filtered(collection, minsize=1, forgive=0):
    """
    Generate partitions that contain at least `minsize` elements per set;
    allow `forgive` missing elements, which can get added in subsequent steps
    """
    if len(collection) == 1:
        yield [ collection ]
        return

    first = collection[0]
    for smaller in partition_filtered(collection[1:], minsize, forgive=forgive+1):
        # insert `first` in each of the subpartition's subsets
        for n, subset in enumerate(smaller):
            candidate = smaller[:n] + [[ first ] + subset]  + smaller[n+1:]
            if conforms(candidate, minsize, forgive):
                yield candidate

        # put `first` in its own subset
        candidate = [ [ first ] ] + smaller
        if conforms(candidate, minsize, forgive):
            yield candidate


import time

t = time.time()
something = list(range(1, int(sys.argv[1])))
v = partition_filtered(something, minsize=2)
x = 0
for p in v:
    p.sort()
    x += p[len(p) // 3][0]
print(x)
print(time.time() - t)
